engine/resources/docker.go

```diff
@@ -0,0 +1,414 @@
+			return true, nil
+		}
+		id = containerList[0].ID // save the id for later
+		// If the image is wrong, and force is true, mark the container for
+		// destruction.
+		if containerList[0].Image != obj.Image && obj.Force {
+			destroy = true
+		}
+		// Otherwise return an error.
+		if containerList[0].Image != obj.Image && !obj.Force {
+			return false, fmt.Errorf("%s exists but has the wrong image: %s", obj.Name(), containerList[0].Image)
+
+		}
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	if obj.State == ContainerStopped { // container exists and should be stopped
+		return false, containerStop(ctx, obj.client, id, nil)
+	}
+
+	if obj.State == ContainerRemoved { // container exists and should be removed
+		if err := containerStop(ctx, obj.client, id, nil); err != nil {
+			return false, err
+		}
+		return false, containerRemove(ctx, obj.client, id, types.ContainerRemoveOptions{})
+	}
+
+	if destroy {
+		if err := containerStop(ctx, obj.client, id, nil); err != nil {
+			return false, err
+		}
+		if err := containerRemove(ctx, obj.client, id, types.ContainerRemoveOptions{}); err != nil {
+			return false, err
+		}
+		containerList = []types.Container{} // zero the list
+	}
+
+	if len(containerList) == 0 { // no container was found
+		_, err = obj.client.ImagePull(ctx, obj.Image, types.ImagePullOptions{})
+		if err != nil {
+			return false, fmt.Errorf("error pulling image: %s", err)
+		}
+
+		containerConfig := &container.Config{
+			Image:        obj.Image,
+			Cmd:          obj.Cmd,
+			Env:          obj.Env,
+			ExposedPorts: make(map[nat.Port]struct{}),
+		}
+
+		hostConfig := &container.HostConfig{
+			PortBindings: make(map[nat.Port][]nat.PortBinding),
+		}
+
+		for k, v := range obj.Ports {
+			for p, q := range v {
+				containerConfig.ExposedPorts[nat.Port(k)] = struct{}{}
+				hostConfig.PortBindings[nat.Port(fmt.Sprintf("%d/%s", p, k))] = []nat.PortBinding{
+					{
+						HostIP:   "0.0.0.0",
+						HostPort: fmt.Sprintf("%d", q),
+					},
+				}
+			}
+		}
+
+		c, err := obj.client.ContainerCreate(ctx, containerConfig, hostConfig, nil, obj.Name())
+		if err != nil {
+			return false, fmt.Errorf("error creating container: %s", err)
+		}
+		id = c.ID
+	}
+
+	return false, containerStart(ctx, obj.client, id, types.ContainerStartOptions{})
+}
+
+// containerStart starts the specified container, and waits for it to start.
+func containerStart(ctx context.Context, client *client.Client, id string, opts types.ContainerStartOptions) error {
+	// Get an events channel for the container we're about to start.
+	eventOpts := types.EventsOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{"container", id}),
+	}
+	eventCh, errCh := client.Events(ctx, eventOpts)
+	// Start the container.
+	if err := client.ContainerStart(ctx, id, opts); err != nil {
+		return fmt.Errorf("error starting container: %s", err)
+	}
+	// Wait for a message on eventChan that says the container has started.
+	select {
```

Comment:
Is it possible that either of these never occurs and we block indefinitely?

Reply:
I don't think so but I'll add a `case <- ctx.Done():` for safety.